tutorials/Land/Soil/Water/hydraulic_functions.jl

```diff
@@ -0,0 +1,352 @@
+# `NoImpedance{FT}`, and for `moisture_factor`, using the type `MoistureIndependent{FT}`.
+
+# In the case where the `moisture_factor = MoistureDependent{FT}()`, either the van
+# Genuchten or Brooks and Corey expression is used based on the type of `hydraulics` model
+# passed.
+
+# The `moisture_factor` for the van Genuchten model is (denoting it as ``K_m``)
+
+# ``
+#  K_m = \sqrt{S_l}[1-(1-S_l^{1/m})^m]^2,
+# ``
+
+# for ``S_l < 1``,
+
+# and for the Brooks and Corey model it is
+
+# ``
+# K_m = S_l^{2M+3},
+# ``
+
+# also for ``S_l<1``. When ``S_l\geq 1``, ``K_m = 1`` for each model.
+
+# One side effect of this flexibility is that `hydraulic_conductivity`
+# requires all the arguments it could possibly need passed to it, which is
+# why here we must supply a value `T` and `θ_i`, even though they are not used
+# in this particular example.
+T = FT(0.0)
+θ_i = FT(0.0)
+
+K =
+    Ksat .*
+    hydraulic_conductivity.(
+        Ref(impedance_choice),
+        Ref(viscosity_choice),
+        Ref(moisture_choice),
+        Ref(hydraulics),
+        Ref(θ_i),
+        Ref(ν),
+        Ref(T),
+        S_l,
+    );
+# Let's also compute `K` when we include the effects of temperature
+# and ice on the hydraulic conductivity.
+# In the cases where a `TemperatureDependentViscosity{FT}` or
+# `IceImpedance{FT}` type is passed, the correct factors are calculated,
+# based on the temperature `T` and volumetric ice fraction `θ_i`.
+# In these cases, the `viscosity_factor`, denoted here as ``K_v``, evaluates as:
+
+# ``
+# K_v = e^{\gamma(T-T_{ref})},
+# ``
+
+# where ``\gamma = 0.0264 \mbox{K}^{-1}`` is an empirical factor, and ``T_{ref} = 288`` K,
+# and the `impedance_factor`, denoted ``K_i``, evaluates as:
+
+# ``
+# K_i = 10^{-\Omega S_i}
+# ``
+
+# where ``\Omega = 7`` is an empirical factor and ``S_i`` is the ratio of the volumetric
+# ice fraction to porosity [5].
+
+viscosity_choice_T = TemperatureDependentViscosity{FT}()
+T = FT(300.0)
+K_T =
+    Ksat .*
+    hydraulic_conductivity.(
+        Ref(impedance_choice),
+        Ref(viscosity_choice_T),
+        Ref(moisture_choice),
+        Ref(hydraulics),
+        Ref(θ_i),
+        Ref(ν),
+        Ref(T),
+        S_l,
+    )
+ice_impedance_I = IceImpedance{FT}()
+θ_i = FT(0.3)
+S_l_accounting_for_ice = FT.(0.01:0.01:0.7) # note that the total volumetric water fraction cannot exceed unity.
+K_i =
+    Ksat .*
+    hydraulic_conductivity.(
+        Ref(ice_impedance_I),
+        Ref(viscosity_choice),
+        Ref(moisture_choice),
+        Ref(hydraulics),
+        Ref(θ_i),
+        Ref(ν),
+        Ref(T),
+        S_l_accounting_for_ice,
+    )
+plot(
+    S_l,
+    log10.(K),
+    xlabel = "effective saturation",
+    ylabel = "Log10(K)",
+    label = "Base case",
+)
+plot!(S_l, log10.(K_T), label = "Temperature Dependent Viscosity")
+plot!(S_l_accounting_for_ice, log10.(K_i), label = "Ice Impedance")
+savefig("./T_ice_K.png")
```

Comment:
```suggestion
savefig("T_ice_K.png")
```
`/` is an OS-specific path separator. Better to always avoid this, and use `joinpath` if we're saving somewhere other than `pwd`.

Reply:
ah, ok! I thought it universally just meant in the current directory. will change.